File: cli/core/tasks.go
// Lista do dia (spec/SPEC.md §4.3) — RF02/RF03.
// Tarefas ativas ordenadas A→B→C (sem prioridade = C); empate por criação.
package core

import "slices"

const maxActiveA = 4

var priorityRank = map[string]int{"A": 0, "B": 1, "C": 2}

func rankOf(priority *string) int {
	if priority == nil {
		return priorityRank["C"]
	}
	return priorityRank[*priority]
}

type Task struct {
	Title    string  `json:"title"`
	ParentID *string `json:"parent_id"`
	DueDate  *string `json:"due_date"`
	Priority *string `json:"priority"`
	Done     bool    `json:"done"`
}

type TasksState struct {
	Tasks   map[string]Task `json:"tasks"`
	DayList []string        `json:"day_list"`
	Alerts  TaskAlerts      `json:"alerts"`
}

type TaskAlerts struct {
	TooManyA bool `json:"too_many_a"`
}

func init() {
	Register("tasks", func(events []Event, _ Params) (any, error) {
		return Tasks(events)
	})
}

func Tasks(events []Event) (TasksState, error) {
	type taskState struct {
		Task
		id      string
		seq     int
		deleted bool
	}
	tasks := map[string]*taskState{}
	var order []*taskState

	for _, event := range Normalize(events) {
		switch event.Type {
		case "task.created":
			pl, err := decodePayload[struct {
				TaskID   string  `json:"task_id"`
				Title    string  `json:"title"`
				ParentID *string `json:"parent_id"`
				DueDate  *string `json:"due_date"`
			}](event)
			if err != nil {
				return TasksState{}, err
			}
			task := &taskState{
				Task: Task{Title: pl.Title, ParentID: pl.ParentID, DueDate: pl.DueDate},
				id:   pl.TaskID,
				seq:  len(order),
			}
			tasks[pl.TaskID] = task
			order = append(order, task)

		case "task.prioritized":
			pl, err := decodePayload[struct {
				TaskID   string `json:"task_id"`
				Priority string `json:"priority"`
			}](event)
			if err != nil {
				return TasksState{}, err
			}
			if task, ok := tasks[pl.TaskID]; ok {
				priority := pl.Priority
				task.Priority = &priority // última repriorização vence
			}

		case "task.edited":
			pl, err := decodePayload[struct {
				TaskID  string  `json:"task_id"`
				Title   *string `json:"title"`
				DueDate *string `json:"due_date"`
			}](event)
			if err != nil {
				return TasksState{}, err
			}
			if task, ok := tasks[pl.TaskID]; ok {
				// last-writer-wins por campo: só os campos presentes mudam
				if pl.Title != nil {
					task.Title = *pl.Title
				}
				if pl.DueDate != nil {
					task.DueDate = pl.DueDate
				}
			}

		case "task.deleted":
			pl, err := decodePayload[struct {
				TaskID string `json:"task_id"`
			}](event)
			if err != nil {
				return TasksState{}, err
			}
			if task, ok := tasks[pl.TaskID]; ok {
				task.deleted = true // tombstone: eventos posteriores são ignorados
				delete(tasks, pl.TaskID)
			}

		case "task.completed":
			pl, err := decodePayload[struct {
				TaskID string `json:"task_id"`
			}](event)
			if err != nil {
				return TasksState{}, err
			}
			if task, ok := tasks[pl.TaskID]; ok {
				task.Done = true
			}
		}
	}

	var active []*taskState
	for _, task := range order {
		if !task.Done && !task.deleted {
			active = append(active, task)
		}
	}

	dayList := slices.Clone(active)
	slices.SortStableFunc(dayList, func(a, b *taskState) int {
		if dr := rankOf(a.Priority) - rankOf(b.Priority); dr != 0 {
			return dr
		}
		return a.seq - b.seq
	})

	activeA := 0
	for _, task := range active {
		if task.Priority != nil && *task.Priority == "A" {
			activeA++
		}
	}

	state := TasksState{
		Tasks:   make(map[string]Task, len(tasks)),
		DayList: []string{},
		Alerts:  TaskAlerts{TooManyA: activeA > maxActiveA},
	}
	for id, task := range tasks {
		state.Tasks[id] = task.Task
	}
	for _, task := range dayList {
		state.DayList = append(state.DayList, task.id)
	}
	return state, nil
}

File: extension/core/tasks.js
// Lista do dia (spec/SPEC.md §4.3) — RF02/RF03.
// Tarefas ativas ordenadas A→B→C (sem prioridade = C); empate por criação.

import { normalize } from "./envelope.js";

const PRIORITY_RANK = { A: 0, B: 1, C: 2 };
const MAX_ACTIVE_A = 4;

function rankOf(priority) {
  return PRIORITY_RANK[priority ?? "C"];
}

export function project(events) {
  const tasks = new Map();
  let seq = 0;

  for (const event of normalize(events)) {
    const { payload } = event;
    if (event.type === "task.created") {
      tasks.set(payload.task_id, {
        title: payload.title,
        parent_id: payload.parent_id ?? null,
        due_date: payload.due_date ?? null,
        priority: null,
        done: false,
        seq: seq++,
      });
    } else if (event.type === "task.prioritized") {
      const task = tasks.get(payload.task_id);
      if (task) task.priority = payload.priority; // última repriorização vence
    } else if (event.type === "task.edited") {
      const task = tasks.get(payload.task_id);
      if (task) {
        // last-writer-wins por campo: só os campos presentes mudam
        if (payload.title !== undefined) task.title = payload.title;
        if (payload.due_date !== undefined) task.due_date = payload.due_date;
      }
    } else if (event.type === "task.deleted") {
      tasks.delete(payload.task_id); // tombstone: eventos posteriores são ignorados
    } else if (event.type === "task.completed") {
      const task = tasks.get(payload.task_id);
      if (task) task.done = true;
    }
  }

  const active = [...tasks.entries()].filter(([, t]) => !t.done);

  const day_list = active
    .slice()
    .sort(([, a], [, b]) => {
      const dr = rankOf(a.priority) - rankOf(b.priority);
      return dr !== 0 ? dr : a.seq - b.seq;
    })
    .map(([id]) => id);

  const activeA = active.filter(([, t]) => t.priority === "A").length;

  const tasksOut = {};
  for (const [id, { title, parent_id, due_date, priority, done }] of tasks) {
    tasksOut[id] = { title, parent_id, due_date, priority, done };
  }

  return {
    tasks: tasksOut,
    day_list,
    alerts: { too_many_a: activeA > MAX_ACTIVE_A },
  };
}
